InnerEye/ML/dataset/scalar_dataset.py

```diff
@@ -28,38 +28,44 @@
 from InnerEye.ML.utils.features_util import FeatureStatistics
 from InnerEye.ML.utils.transforms import Compose3D, Transform3D
 
-
 T = TypeVar('T', bound=ScalarDataSource)
 
 
-def extract_label_classification(label_string: Union[str, float], sample_id: str) -> Union[float, int]:
+def extract_label_classification(label_string: str, sample_id: str, num_classes: int) -> List[float]:
     """
     Converts a string from a dataset.csv file that contains a model's label to a scalar.
     The function maps ["1", "true", "yes"] to 1, ["0", "false", "no"] to 0.
     If the entry in the CSV file was missing (no string given at all), it returns math.nan.
     :param label_string: The value of the label as read from CSV via a DataFrame.
     :param sample_id: The sample ID where this label was read from. This is only used for creating error messages.
+    :param num_classes: Number of expected classes
     :return:
     """
-    if isinstance(label_string, float):
-        if math.isnan(label_string):
-            # When loading a dataframe with dtype=str, missing values can be encoded as NaN, and get into here.
-            return label_string
-        else:
-            raise ValueError(f"Subject {sample_id}: Unexpected float input {label_string} - did you read the "
-                             f"dataframe column as a string?")
+
+    if '|' in label_string or label_string.isdigit():
+        if num_classes == 1:
+            if label_string in ["0", "1"]:
+                return [float(label_string)]
+            else:
+                raise ValueError(f"Subject {sample_id}: Label string not recognized: '{label_string}'")
+        classes = [int(a) for a in label_string.split('|')]
+        one_hot_array = np.zeros(num_classes, dtype=np.float)
+        one_hot_array[classes] = 1.0
+        return one_hot_array
+
     if label_string:
         label_lower = label_string.lower()
-        if label_lower in ["1", "true", "yes"]:
-            return 1
-        if label_lower in ["0", "false", "no"]:
-            return 0
+        if label_lower in ["true", "yes"]:
+            return [1.0]
+        if label_lower in ["false", "no"]:
+            return [0.0]
+
         raise ValueError(f"Subject {sample_id}: Label string not recognized: '{label_string}'")
     else:
-        return math.nan
+        return [math.nan]
 
 
-def extract_label_regression(label_string: Union[str, float], sample_id: str) -> Union[float, int]:
+def extract_label_regression(label_string: Union[str, float], sample_id: str, num_classes: int) -> List[float]:
```

Comment:
why do we have num_classes as an argument here at all? it is not used apart from doing a check?

Reply:
There some function that gets created expecting this interface :/